linux/src/threadpool_linux.c

```diff
@@ -531,7 +521,30 @@ int threadpool_schedule_work(THANDLE(THREADPOOL) threadpool, THREADPOOL_WORK_FUN
     return result;
 }
 
-int threadpool_timer_start(THANDLE(THREADPOOL) threadpool, uint32_t start_delay_ms, uint32_t timer_period_ms, THREADPOOL_WORK_FUNCTION work_function, void* work_function_ctx, TIMER_INSTANCE_HANDLE* timer_handle)
+static void threadpool_timer_dispose(TIMER * timer)
+{
+    /* Codes_SRS_THREADPOOL_LINUX_07_071: [ threadpool_timer_cancel shall call timer_delete to destroy the ongoing timers. ]*/
+    if (timer_delete(timer->time_id) != 0)
+    {
+        LogErrorNo("Failure calling timer_delete.");
+    }
+    else
+    {
+        // Do Nothing
+    }
+    srw_lock_ll_deinit(&timer->timer_lock);
+    // Even though timer_delete does cancel any events, there is a small window where an event was triggered just before and a thread is being created.
+    // So the callback can still execute after the timer_delete call. A small wait here keeps the timer instance alive long enough for
+    // the callback to use it if we hit this window.
+
+    // This doesn't fix the problem, this won't 100% guarantee the thread will be executed before we free the timer data. There are a couple of
+    // options to fix the problem, but the way that fits this threadpool model best would be to keep a pool of timers in the threadpool,
+    // and manage each state as in-use or not. That way the timer data allocation is not dependent on threadpool_timer_start/destroy.
+    // To reproduce this problem, create a large number of threads that delete the timer at the same time it is due to expire and remove the pause
+    // above.
+}
```

Comment:
This mentions a small wait that does not exist any longer. I think this comment is now obsolete?

Reply:
deleted